doc/Dump-Utility.md

```diff
@@ -0,0 +1,823 @@
+# Dump Utility for Easy Debugging #
+#### Rev 1.0
+
+# Table of Contents
+  * [Revision](#revision)
+  * [About this Manual](#about-this-manual)
+  * [Definitions/Abbreviation](#definitionsabbreviations)
+  * [Overview](#overview)
+  * [1. Requirements](#1-requirements)
+      * [1.1 Intended Usage](#11-intended-usage)
+      * [1.2 Customization Options](#12-customization-options)
+      * [1.3 Extensibility](#13-extensibility)
+      * [1.4 Directory Structure](#14-directory-structure)
+  * [2. Design](#2-design)
+      * [2.0 The Executor Class](#20-the-executor-class)
+      * [2.1 Adding new Modules](#21-adding-new-modules)
+      * [2.2 Module Internals](#22-module-internals)
+      * [2.3 Helper Methods Available](#23-helper-methods-available)
+      * [2.4 Match Infrastructure](#24-Match-Infrastructure)
+      * [2.5 MatchRequest Examples](#25-matchrequest-examples)
+  * [3. Unit Tests](#3-unit-tests)
+  * [4. TechSupport](#4-techsupport)
+
+### Revision  
+
+| Rev |     Date    |       Author       | Change Description          |
+|:---:|:-----------:|:-------------------------|:----------------------|
+| 1.0 | 05/28/2021  | Vivek Reddy Karri        | Initial version       |
+
+## About this Manual
+This document describes the details of a dump cli utility which collects and dumps the redis state for a given feature/module.
+
+## Definitions/Abbreviations
+###### Table 1: Abbreviations
+| Abbreviation | Description                                                  |
+| ------------ | ------------------------------------------------------------ |
+| APP DB       | Application Database                                         |
+| ASIC DB      | ASIC Database                                                |
+| CONF DB      | Configuration Database                                       |
+
+
+
+In this document, the term '**redis state**' refers to the intermediate state of a given feature present across all the Redis DB's
+
+## Overview
+In SONiC, there usually exists a set of tables related/relevant to a particular module. All of these have to be looked at to confirm whether  any configuration update is properly applied and propagated.
+
+The task of debugging quickly becomes tedious because currently, there is no utility which does print a unified view of the redis-state.
+This is the problem which is addressed by this dump utility.
+
+This utility provides the base infrastructure and guidelines to make is easy for the developers to extend and support the utility for different modules.
+
+## **1 Requirements**
+
+### 1.1 Intended Usage
+
+Note: Currently, `state` is the only subcommand under `dump`.
```

Comment:
Will 'state' be confusing to understand.
For example, if we say dump "state", will use expect to dump the "state_db"? 
Will user think that, dump could either dump state or dump config.
What is the intention of adding sub command under dump? Is the idea to use dump for any thing else in future? is there an example where dump could be used for something else in future?

Reply:
Yeah, that was indeed the idea, if we choose to use the dump command for something else apart from fetching just redis-state. For example, Maybe to fetch data from sdk dumps, etc.

As for the state subcommand,  the help should make it clear.

```
root@sonic# dump state --help
Usage: dump state [OPTIONS] MODULE IDENTIFIER

  Dump the redis-state of the identifier for the module specified

Options:
  -s, --show            Display Modules Available
  -d, --db TEXT         Only dump from these Databases
  -t, --table           Print in tabular format  [default: False]
  -k, --key-map         Only fetch the keys matched, don't extract field-value dumps  [default: False]
  -v, --verbose         Prints any intermediate output to stdout useful for dev & troubleshooting  [default: False]
  -n, --namespace TEXT  Dump the redis-state for this namespace.  [default: DEFAULT_NAMESPACE]
  --help                Show this message and exit.
```

Reply:
Ok, The help sentence should be modified to convey that the CLI is not just to dump redis-state but also can dump from a CONFIG_FILE.
Also, is there any help in the command that tells the user that -d/--db option can take db name or CONFIG_FILE?

Reply:
1) Yeah, will add that
2) Currently, there is no way to notify the user about the DB/CONFIG_FILE's that the particular module will be looking, directly from the CLI. Can add this to the roadmap for next revision though.

Reply:
Ok, It is better to update the help message for --db sub command, telling "Dump from specified database or config_file".

Reply:
Updated them in the HLD.

Reply:
Help message of "dump state" could be shortened as:
 "Dump the current state of the identifier for the specified module from Redis DB or CONFIG_FILE"

Reply:
Done